x/market/handler/endblock_test.go

```diff
@@ -0,0 +1,84 @@
+package handler
+
+import (
+	"fmt"
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/ovrclk/akash/testutil"
+	"github.com/ovrclk/akash/x/market/types"
+)
+
+type winnerTest struct {
+	bids      []types.Bid
+	expWinner types.Bid
+}
+
+func TestBidWinner(t *testing.T) {
+	winBid0 := createBid(t, uint64(20), int64(5))
+
+	bid0 := createBid(t, uint64(15), int64(3))
+	bid1 := createBid(t, uint64(20), int64(5))
+	bid2 := createBid(t, uint64(19), int64(5))
+	bid3 := createBid(t, uint64(20), int64(4))
+	bid4 := createBid(t, uint64(20), int64(5))
+	bid5 := createBid(t, uint64(19), int64(5))
+	bid6 := createBid(t, uint64(18), int64(5))
+	bid7 := createBid(t, uint64(20), int64(8))
+	bid8 := createBid(t, uint64(20), int64(6))
+
+	var winnerTests []winnerTest = []winnerTest{
+		{
+			bids:      []types.Bid{},
+			expWinner: types.Bid{},
+		},
+		{
+			bids:      []types.Bid{winBid0},
+			expWinner: winBid0,
+		},
+		{
+			bids:      []types.Bid{winBid0, bid7},
+			expWinner: winBid0,
+		},
+		{
+			bids:      []types.Bid{winBid0, bid3},
+			expWinner: bid3,
+		},
+		{
+			bids:      []types.Bid{bid1, bid0, winBid0, bid2, bid3, bid4},
+			expWinner: bid0,
+		},
+		{
+			bids:      []types.Bid{bid5, bid6, bid7, bid8},
+			expWinner: bid6,
+		},
+	}
+
+	for i, test := range winnerTests {
+		f := func(_ *testing.T, test winnerTest) func(t *testing.T) {
+			return func(t *testing.T) {
+				winner := pickBidWinner(test.bids)
+				if !winner.Equals(test.expWinner.ID()) {
+					t.Errorf("unexpected winner: %#v\n%q : %v", winner, types.BidIDString(winner.BidID), winner.Price)
+				}
+			}
+		}
```

Comment:
This is making tests too complicated and difficult to read.  I really think it's over-doing it.

I generally add a message to table-based test elements to use in the assertions.  Something to consider:

```go
tests := []struct{
  name string
  val    int
  expect int
}{
  {
     name: "zero-property",
     val: 0,
     expect: 0,
  },
}
for _, test := tests {
  assert.Equal(t,test.expect,foo(test.val),test.name)
}
```

That was before `t.Run()`, and I get that you're hitting lint and semantic issues with it, but there has to be a better way around it.

```go
for i := range winnerTests {
  t.Run(strconv.Itoa(i),func(t *testing.T) {
    assert.Equals(t,winnerTests[i].expWinner.ID(),pickBidWinner(winnerTests[i].bids))
  })

Reply:
I mean, is it right?